### PR TITLE
fix(files): drag leave detection on safari

### LINKED ISSUE #43135: [Bug]: NC28.0.2rc4 drag-and-drop still broken on Safari
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

NC28 breaks drag-and-drop at least on Chrome and Safari.
In NC28.0.2rc4, drag-and-drop is fixed on Chrome but not on Safari.

Related issues for Chrome: https://github.com/nextcloud/server/pull/42409
https://github.com/nextcloud/server/pull/42366

We need a fix for Safari as well.

### Steps to reproduce

1. Use Safari
2. Try to upload files or folders with drag-and-drop
3. It does not work

### Expected behavior

Uploading files or folders with drag-and-drop on Safari should work.

### Installation method

Community Manual installation with Archive

### Nextcloud Server version

28

### Operating system

Arch Linux for server, MacOS Sonoma for client.

### PHP engine version

PHP 8.3

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Upgraded to a MAJOR version (ex. 22 to 23)

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.vip-sourcing.ch"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "28.0.2.3",
        "overwrite.cli.url": "https:\/\/cloud.vip-sourcing.ch",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "ssl",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauth": 1,
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "loglevel": 0,
        "default_phone_region": "CH",
        "filelocking.enabled": "true",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "dbindex": 0,
            "password": "***REMOVED SENSITIVE VALUE***",
            "timeout": 1.5
        },
        "app_install_overwrite": [
            "mail_roundcube"
        ],
        "theme": "",
        "updater.release.channel": "beta"
    }
}

### List of activated Apps

Enabled:
  - activity: 2.20.0
  - calendar: 4.6.4
  - circles: 28.0.0-dev
  - cloud_federation_api: 1.11.0
  - comments: 1.18.0
  - contacts: 5.5.1
  - contactsinteraction: 1.9.0
  - dashboard: 7.8.0
  - dav: 1.29.1
  - deck: 1.12.1
  - federatedfilesharing: 1.18.0
  - federation: 1.18.0
  - files: 2.0.0
  - files_pdfviewer: 2.9.0
  - files_reminders: 1.1.0
  - files_sharing: 1.20.0
  - files_trashbin: 1.18.0
  - files_versions: 1.21.0
  - firstrunwizard: 2.17.0
  - impersonate: 1.15.0
  - integration_google: 2.1.0
  - logreader: 2.13.0
  - lookup_server_connector: 1.16.0
  - mail: 3.5.5
  - mail_roundcube: 1.0.3
  - nextcloud_announcements: 1.17.0
  - notifications: 2.16.0
  - oauth2: 1.16.3
  - password_policy: 1.18.0
  - photos: 2.4.0
  - privacy: 1.12.0
  - provisioning_api: 1.18.0
  - recommendations: 2.0.0
  - related_resources: 1.3.0
  - richdocuments: 8.3.1
  - serverinfo: 1.18.0
  - settings: 1.10.1
  - sharebymail: 1.18.0
  - support: 1.11.0
  - survey_client: 1.16.0
  - systemtags: 1.18.0
  - text: 3.9.1
  - theming: 2.3.0
  - twofactor_backupcodes: 1.17.0
  - updatenotification: 1.18.0
  - user_status: 1.8.1
  - viewer: 2.2.0
  - weather_status: 1.8.0
  - workflowengine: 2.10.0
Disabled:
  - admin_audit: 1.18.0
  - bruteforcesettings: 2.8.0
  - encryption: 2.16.0
  - files_external: 1.20.0
  - files_rightclick: 0.15.1 (installed 1.6.0)
  - onlyoffice: 9.0.0 (installed 9.0.0)
  - suspicious_login: 6.0.0
  - twofactor_totp: 10.0.0-beta.2
  - user_ldap: 1.19.0

### Nextcloud Signing status

No errors have been found.

### Nextcloud Logs

None with LogLevel 1 set.

### Additional info

Safari console log:
```
[Log] JQMIGRATE: Migrate is installed, version 3.4.1 (core-main.js, line 2)
[Info] [INFO] activity: Activity API registered (logger-a53a57b5.mjs, line 6)
Object

app: "activity"

level: 1

uid: "gdhautefeuille"

Object Prototype
[Debug] Unified search initialized! (core-unified-search.js, line 2)
[Debug] OCA.Files.Settings initialized (files-main.js, line 2)
[Debug] OCA.Files.Sidebar initialized (files-sidebar.js, line 2)
[Error] TypeError: undefined is not an object (evaluating 'OCA.Files.FileList.prototype')
	(anonymous function) (circles.files.list.js:42)
	Global Code (circles.files.list.js:306)
[Debug] OCA.Comments.View initialized (comments-comments-app.js, line 2)
[Warning] [WARN] viewer: Some mimes were ignored because they are not enabled in the server previews config – Object (viewer-main.js, line 2)
Object
[Debug] OCA.Viewer initialized (viewer-main.js, line 2)
[Debug] OCA.Sharing.ShareSearch initialized (files_sharing-files_sharing_tab.js, line 2)
[Debug] OCA.Sharing.ExternalLinkActions initialized (files_sharing-files_sharing_tab.js, line 2)
[Debug] OCA.Sharing.ExternalShareActions initialized (files_sharing-files_sharing_tab.js, line 2)
[Info] [INFO] viewer: 6 viewer handlers registered – Object (viewer-main.js, line 2)
Object
[Info] session heartbeat polling started (core-main.js, line 2)
[Info] [INFO] comments: Comments plugin registered for Activity sidebar action – Object (core-common.js, line 2)
Object
[Debug] Registering notifications container as a menu (notifications-node_modules_moment_locale_sync_recursive_-src_NotificationsApp_vue-data_image_svg_xml_base64-ab95a3.js, line 2)
[Info] Notifications permissions not yet requested (notifications-node_modules_moment_locale_sync_recursive_-src_NotificationsApp_vue-data_image_svg_xml_base64-ab95a3.js, line 2)
[Debug] Polling interval updated to 30000 (notifications-node_modules_moment_locale_sync_recursive_-src_NotificationsApp_vue-data_image_svg_xml_base64-ab95a3.js, line 2)
[Debug] Started background fetcher as session_keepalive is enabled (notifications-node_modules_moment_locale_sync_recursive_-src_NotificationsApp_vue-data_image_svg_xml_base64-ab95a3.js, line 2)
[Debug] Search providers – Array (11) (core-unified-search.js, line 2)
Array (11)
[Debug] Contacts – Array (9) (core-unified-search.js, line 2)
Array (9)
[Debug] Got notification data, restoring default polling interval. (notifications-node_modules_moment_locale_sync_recursive_-src_NotificationsApp_vue-data_image_svg_xml_base64-ab95a3.js, line 2)
[Debug] Mounted – "workspace" (files-main.js, line 2)
[Debug] Mounted – "recommendations" (files-main.js, line 2)
[Error] [ERROR] files: Header workspace already registered
Object

app: "files"

header: w {_header: Object}

level: 1

uid: "gdhautefeuille"

Object Prototype
	value (text-files.js:2:11066)
	value (text-files.js:2:11536)
	j (text-files.js:2:209945)
	(anonymous function) (text-files.js:2:258898)
```

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix https://github.com/nextcloud/server/issues/43135

| Before | After |
|:---------:|:------:|
|old|new|

### CODE REVIEW

File: apps/files/src/components/DragAndDropNotice.vue [hunk 1]
```diff
@@ -126,7 +126,7 @@ export default defineComponent({
 			// only when we're leaving the current element
 			// Avoid flickering
 			const currentTarget = event.currentTarget as HTMLElement
-			if (currentTarget?.contains(event.relatedTarget as HTMLElement)) {
+			if (currentTarget?.contains((event.relatedTarget || event.target) as HTMLElement)) {
```
Comment (susnux) [inline apps/files/src/components/DragAndDropNotice.vue:129]:
```suggestion
			if (currentTarget?.contains((event.relatedTarget ?? event.target) as HTMLElement)) {
```

To make it clearer that `relatedTarget` is null?

**Review by susnux — APPROVED** [review]
I can not verify this fixes the issue (I have no access to Apple devices), but the code makes sense as it seems to be related to this long standing WebKit bug where `relatedTarget` is unset: https://bugs.webkit.org/show_bug.cgi?id=66547

(Feels like Safari is the new IE).

**Review by nfebe — APPROVED** [review]

**Comment by skjnldsv** [comment]
> Feels like Safari is the new IE

You and me both 😅🥲

**Comment by susnux** [comment]
> compile rebase-amend

Thats new? Nice!

**Pull request merged**